File: Dart/1-Arrays&Hashing/Problems/3-ContainsDuplicate/ContainsDuplicatePro.dart
import 'dart:collection';

void main() {
  List<int> nums = [1, 2, 3, 4, 5, 1];
  print(containsDuplicateHashSet(nums)); // Output: true
  print(containsDuplicateHashMap(nums)); // Output: true
  containsDuplicateStream(nums).then((result) => print(result)); // Output: true
  print(containsDuplicateSQL(nums)); // Simulated SQL check
  print(containsDuplicateBloomFilter(nums)); // Simulated Bloom filter
}

// Approach 1: Using HashSet (Best for general use cases)
// Time Complexity: O(n), Space Complexity: O(n)
bool containsDuplicateHashSet(List<int> nums) {
  Set<int> seen = HashSet();
  for (int num in nums) {
    if (!seen.add(num)) {
      return true;
    }
  }
  return false;
}

// Approach 2: Using HashMap (Counts frequency of elements)
// Time Complexity: O(n), Space Complexity: O(n)
bool containsDuplicateHashMap(List<int> nums) {
  Map<int, int> frequency = HashMap();
  for (int num in nums) {
    if (frequency.containsKey(num)) {
      return true;
    }
    frequency[num] = 1;
  }
  return false;
}

// Approach 3: Using Dart Streams (For reactive programming)
// Time Complexity: O(n), Space Complexity: O(n)
Future<bool> containsDuplicateStream(List<int> nums) async {
  Set<int> seen = HashSet();
  return await Stream.fromIterable(nums).any((num) {
    if (seen.contains(num)) return true;
    seen.add(num);
    return false;
  });
}

// Approach 4: Using a Database Query (SQL-based approach for large datasets)
// Simulating a SQL query that checks for duplicates in a database
// Time Complexity: O(n) (Assuming indexed search in DB)
bool containsDuplicateSQL(List<int> nums) {
  // In real applications, replace this with an actual database query
  return nums.toSet().length != nums.length;
}

// Approach 5: Using a Bloom Filter (For large-scale distributed systems)
// Simulating a Bloom Filter (Probabilistic duplicate detection)
bool containsDuplicateBloomFilter(List<int> nums) {
  Set<int> seen = HashSet(); // Replace with actual Bloom filter in production
  for (int num in nums) {
    if (seen.contains(num)) {
      return true; // In real-world, may return false positives
    }
    seen.add(num);
  }
  return false;
}

File: Python/1-Arrays&Hashing/Problems/3-ContainsDuplicate/ContainsDuplicatePro.py
"""
Problem: Contains Duplicate (Advanced Real-World Implementations)
Given an integer array nums, return true if any value appears at least twice in the array, and false if every element is distinct.

Approaches:
1. Database Query Optimization (SQL-based detection)
2. Distributed Processing (Spark DataFrame)
3. Set-Based Memory-Efficient Duplicate Detection (Replacing Bloom Filter)
4. Streaming Data Processing (Handling Large Data Streams)
5. Cryptographic Hashing (Hash Collisions for Large Datasets)
6. Parallel Processing with Multiprocessing (Optimized for Large Lists)
"""

def contains_duplicate_sql_query(database_connection, table_name):
    """
    Approach 1: Database Query Optimization
    Use SQL to detect duplicates in a dataset stored in a database.
    Time Complexity: O(n)
    Space Complexity: O(1) (operates within the database)
    """
    query = f"SELECT COUNT(*) FROM (SELECT column_name, COUNT(*) as cnt FROM {table_name} GROUP BY column_name HAVING cnt > 1) AS duplicates;"
    cursor = database_connection.cursor()
    cursor.execute(query)
    return cursor.fetchone()[0] > 0

def contains_duplicate_spark(df):
    """
    Approach 2: Distributed Processing (Apache Spark)
    Uses PySpark to detect duplicates efficiently in large datasets.
    Time Complexity: O(n)
    Space Complexity: O(n) (distributed across nodes)
    """
    from pyspark.sql import functions as F
    duplicate_count = df.groupBy("column_name").count().filter(F.col("count") > 1).count()
    return duplicate_count > 0

def contains_duplicate_set(nums):
    """
    Approach 3: Set-Based Memory-Efficient Duplicate Detection
    Uses a simple set to efficiently track seen numbers.
    Time Complexity: O(n)
    Space Complexity: O(n) (worst case, storing all elements)
    """
    seen = set()
    for num in nums:
        if num in seen:
            return True
        seen.add(num)
    return False

def contains_duplicate_streaming(stream):
    """
    Approach 4: Streaming Data Processing
    Handles large real-time data streams efficiently.
    Time Complexity: O(n)
    Space Complexity: O(n) (for stored elements)
    """
    seen = set()
    for num in stream:
        if num in seen:
            return True
        seen.add(num)
    return False

def contains_duplicate_hashing(nums):
    """
    Approach 5: Cryptographic Hashing
    Uses hashing to detect duplicates in large datasets securely.
    Time Complexity: O(n)
    Space Complexity: O(n)
    """
    from hashlib import sha256
    seen = set()
    for num in nums:
        hashed_value = sha256(str(num).encode()).hexdigest()
        if hashed_value in seen:
            return True
        seen.add(hashed_value)
    return False

def chunk_counter(chunk):
    """
    Helper function for multiprocessing.
    """
    from collections import Counter
    return Counter(chunk)

def contains_duplicate_parallel(nums):
    """
    Approach 6: Parallel Processing with Multiprocessing
    Uses multiprocessing to speed up duplicate detection in large lists.
    Time Complexity: O(n)
    Space Complexity: O(n)
    """
    from collections import Counter
    from multiprocessing import Pool, cpu_count

    if not nums:
        return False

    num_workers = min(cpu_count(), len(nums))
    chunk_size = max(1, len(nums) // num_workers)
    chunks = [nums[i * chunk_size:(i + 1) * chunk_size] for i in range(num_workers)]

    with Pool(num_workers) as pool:
        counts = sum(pool.map(chunk_counter, chunks), Counter())

    return any(count > 1 for count in counts.values())

# Example usage (dummy database connection for SQL approach)
# db_connection = setup_database_connection()
# print("Database Query:", contains_duplicate_sql_query(db_connection, "table_name"))

nums = [1, 2, 3, 4, 5, 6, 7, 8, 9, 1]
print("Set-Based Detection:", contains_duplicate_set(nums))
print("Streaming Processing:", contains_duplicate_streaming(iter(nums)))
print("Cryptographic Hashing:", contains_duplicate_hashing(nums))
print("Parallel Processing:", contains_duplicate_parallel(nums))
